### PR TITLE
IOS-7873 Sheet improvements and tests

### PR DESCRIPTION
**Pull request by jmbrocal**
Hi folks!

This PR contains

- Improve panning animation
- Improve the API so the delegate, transition are hidden to the user.
- Bump min iOS version to 13 (cleaning code)
- Fix issues related with dynamic text type
- Fix the maximum height in which the component can take.

What did you do?

The major change of this PR is related with the panning of the component. The goal is improving the panning animation
while simplifying the code. My proposal is similar to the previous implementation: use a UIPresentationController to show a modal view controller with a custom animation but I based the implementation in UIViewPropertyAnimator.

I use two animators: a top offset animator and a height animator:
- offset animator: mainly used for the panning animation when the view controller is dismissing
- height animator: used for simulate the bouncing effect.

This PR is a major update because now the UIPresentationController and UIViewControllerTransitioningDelegate are private, the user of the component presents the component as any other viewcontroller.

I also fixed ui issues when the user has active the dynamic text type setting, or I almost fixed all cases becuase there are two cases in which I cannot fix. The issue is in FontToolkit and how we are calculating the font size considering the dynamic text type setting.

In [FontStyle](https://github.com/Telefonica/mistica-ios/blob/427d1dd1e5b5b8cb147fcf03dbecb6cc40349348/Sources/MisticaCommon/Fonts/FontStyle.swift#L54) we are manually adapting the font size of the textPresset considering the dynamic text type settings but the resolution for the cases `extraExtraExtraLarge` and `accessibilityMedium` are wrong. I was investigating from where did we get those number but I have not luck.

Investigating what Apple offers for scaling a font based on the dynamic text size, I found that in iOS 11 Apple releases a class [UIFontMetrics](https://developer.apple.com/documentation/uikit/uifontmetrics). I created a ticket for fixing the issue [IOS-7873](https://jira.tid.es/browse/IOS-7873). For completing the ticket we need help from design core. They are aware of the problem and needs. As soon as they have the info that we need, they will update the ticket.

I also bumped the min iOS version to 13 (it is not related with the PR, just boyscout job) and fixed the maximum height of the component.

-You can test the changes using this version of the catalog-

The actions are not working (lifecycle is working on it), in the meanwhile you can checkout the branch to test the changes

### CODE REVIEW

File: Sources/Mistica/Utils/Extensions/UIImageView+URL.swift [hunk 1]
```diff
@@ -11,46 +11,63 @@ import UIKit
...
+            darkImageData = try? Data(contentsOf: urlForDarkMode)
         }
+
+        completion(lightImageData, darkImageData)
     }
 
-    func updateImage(imageData: Data, renderAsTemplate: Bool = false) {
-        DispatchQueue.main.async { [weak self] in
-            var image = UIImage(data: imageData)
+    func updateImage(imageData: Data, darkImageData: Data?, renderAsTemplate: Bool = false) {
```
Comment (amegias) [inline Sources/Mistica/Utils/Extensions/UIImageView+URL.swift:54]: This method must be called in main thread
Captura de Pantalla 2022-11-11 a las 12 05 51

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+        heightAnimator = createHeightAnimator(
+            animating: presentedView, from: presentedView.frame.height
+        )
+    }
+
+    func move(_ presentedView: UIView, using translation: CGFloat) {
+        let progress = translation / presentedView.frame.height
+
+        heightAnimator?.fractionComplete = progress * -1
```
Comment (amegias) [inline Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift:46]:
```suggestion
        heightAnimator?.fractionComplete = -progress
```

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    func stop(
+        moving presentedView: UIView,
+        at translation: CGFloat,
+        with velocity: CGPoint
+    ) {
+        let progress = translation / presentedView.frame.height
+
+        heightAnimator?.fractionComplete = progress * -1
```
Comment (amegias) [inline Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift:59]:
```suggestion
        heightAnimator?.fractionComplete = -progress
```

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+//
+//  Copyright © Telefonica. All rights reserved.
+//
+
+import UIKit
+
+final class BottomSheetInteractiveDismissalTransition: NSObject {
+    private let maxBouncingHeight: CGFloat = 250
+    private let animationDuration: CGFloat = UIView.defaultAnimationDuration
+    private let animationCurve: UIView.AnimationCurve = .easeOut
```
Comment (amegias) [inline Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift:14]:
```suggestion
    private enum Constants {
        static let maxBouncingHeight: CGFloat = 250
        static let animationDuration: CGFloat = UIView.defaultAnimationDuration
        static let animationCurve: UIView.AnimationCurve = .easeOut
    }
```

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        to offset: CGFloat,
+        transitionContext: UIViewControllerContextTransitioning
+    ) -> UIViewPropertyAnimator {
+        let propertyAnimator = UIViewPropertyAnimator(
+            duration: animationDuration,
+            curve: animationCurve
+        )
+
+        propertyAnimator.addAnimations {
+            print(offset)
```
Comment (amegias) [inline Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift:210]:
```suggestion
```

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        with velocity: CGPoint
+    ) {
+        let progress = translation / presentedView.frame.height
+
+        heightAnimator?.fractionComplete = progress * -1
+        offsetAnimator?.fractionComplete = progress
+
+        transitionContext?.updateInteractiveTransition(progress)
+
+        let finishDismiss = velocity.y > 100 || progress > 0.45
```
Comment (amegias) [inline Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift:64]: I would move those constants to a Constants enum to assign names to them

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetTransitioningDelegate.swift [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    func interactionControllerForDismissal(
+        using animator: UIViewControllerAnimatedTransitioning
+    ) -> UIViewControllerInteractiveTransitioning? {
+        animator as? BottomSheetInteractiveDismissalTransition
+    }
+}
+
+// MARK: BottomSheetInteractiveDismissalTransition
```
Comment (amegias) [inline Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetTransitioningDelegate.swift:50]:
```suggestion
```

**Review by amegias — APPROVED** [review]
Good work!!!

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift (hunk 2 above)
Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift:46]: Sorry it is not the same, heightAnimator is a reverse animator so we need to multiply the progress for -1

File: Sources/Mistica/Utils/Extensions/UIImageView+URL.swift (hunk 1 above)
Comment (jmbrocal) [reply Sources/Mistica/Utils/Extensions/UIImageView+URL.swift:54]: Good catch, I forget to call it from main thread, Fixed!

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/DimmedView.swift [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            }
+        }
+    }
+
+    /// The closure to be executed when a tap occurs
+    var didTap: ((_ recognizer: UIGestureRecognizer) -> Void)?
+
+    private lazy var tapGesture: UIGestureRecognizer = {
+        UITapGestureRecognizer(target: self, action: #selector(didTapView))
+    }()
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/DimmedView.swift:33]:
```suggestion
    private lazy var tapGesture = UITapGestureRecognizer(target: self, action: #selector(didTapView))
```

File: Sources/Mistica/Utils/Extensions/UIImageView+URL.swift [hunk 9]
```diff
@@ -11,46 +11,66 @@ import UIKit
...
+                urlForDarkMode: urlForDarkMode,
+                completion: { [weak self] lightImageData, darkImageData in
+                    guard let lightImageData else { return }
+                    
+                    DispatchQueue.main.async {
+                        self?.updateImage(imageData: lightImageData, darkImageData: darkImageData)
+                    }
+                }
+            )
         }
```
Comment (jmpg93) [inline Sources/Mistica/Utils/Extensions/UIImageView+URL.swift:37]: Wait, what? What is the difference?

File: Package.swift [hunk 10]
```diff
@@ -5,7 +5,7 @@ import PackageDescription
 let package = Package(
     name: "Mistica",
     platforms: [
-        .iOS(.v12)
+        .iOS(.v13)
```
Comment (jmpg93) [inline Package.swift:8]: Can you remove all the `@available(iOS 13, *)`?

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        }
+
+        let fractionComplete = offsetAnimator?.fractionComplete ?? 0
+
+        offsetAnimator?.stopAnimation(true)
+
+        let offset = presentedView.frame.height
+        let offsetAnimator = createOffsetAnimator(
+            animating: presentedView,
+            to: offset,
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift:150]: I'm seeing a weird dismissal animation but this seems ok...

https://user-images.githubusercontent.com/5618731/201701753-4306872e-173b-4083-aa28-f783a4ee7fdb.mp4

Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift:150]: Please, disable slow animations CMD + T or Debug > Slow Animatins

File: Package.swift (hunk 10 above)
Comment (jmbrocal) [reply Package.swift:8]: Sure

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/DimmedView.swift (hunk 8 above)
Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/DimmedView.swift:33]: oki

File: Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift (hunk 11 above)
Comment (jmpg93) [reply Sources/Mistica/Components/Sheet/View/BottomSheetPresentation/BottomSheetInteractiveDismissalTransition.swift:150]: What do you mean? I enabled slow animations to make the video. The sheet is removed from the hierarchy before dismissing completely.

**Review by jmpg93 — APPROVED** [review]

**Pull request merged**